### PR TITLE
feat(web): specialized tool cards for session stream

### LINKED ISSUE #935: Specialized tool cards for session event stream
## Problem

Tool calls in the session stream render as raw JSON blobs. Every tool — file reads, edits, shell commands, searches — gets the same generic treatment: `> ToolName` followed by `JSON.stringify(args, null, 2)` in a `<pre>` block. Tool results dump raw content (escaped JSON strings, entire file contents, diffs-as-strings) with no formatting.

This is ugly and hard to scan. A user watching an agent work should be able to glance at the stream and immediately understand what's happening without parsing JSON in their head.

### Screenshots (current state)

Three agents ran the same task ("update README.md") — here's what each looks like today:

**Claude Code** — `Read` and `Edit` shown as raw JSON args:
Claude Code session
- `Read` shows `{ "file_path": "C:/Users/nickp/src/grackle4/README.md" }` — why not just show the filename?
- `Edit` dumps `old_string` / `new_string` as escaped JSON with `\n` everywhere — should be a diff

**Copilot** — `view` result dumps the entire 297-line README as a single escaped JSON string:
Copilot session
- Tool result has a `detailedContent` field containing a proper unified diff, but it's rendered as raw escaped JSON
- `report_intent` is metadata noise — shouldn't look like a code operation

**Codex** — `command_execution` shows the command with quadruple-escaped backslashes:
Codex session
- Command: `"\"C:\\\\Program Files\\\\PowerShell\\\\7\\\\pwsh.exe\" -Command 'Get-Content -Path README.md'"` — unreadable
- Output is better (rendered as text) but has no exit-code highlighting or collapsible preview

### Additionally: mock data gap

The `MOCK_STREAM_SCENARIOS` in `mockData.ts` (used by `?mock` mode / demo site) contain **zero** tool_use or tool_result events — only `output` text events. The demo site doesn't showcase tool cards at all.

---

### Principle: the collapsed state tells you what happened

Each tool card should communicate its action at a glance in **one line**, without expanding. The expanded state provides detail for when you want to inspect.

### Tool categories

Classify tools by matching the tool name (case-insensitive, across all runtimes):

| Category | Tool names | Accent color |
|----------|-----------|-------------|
| **File Read** | `Read`, `view` | blue (existing) |
| **File Edit** | `Edit`, `edit`, `file_change` | orange |
| **File Write** | `Write` | green |
| **Shell** | `Bash`, `command_execution` | neutral (terminal) |
| **Search** | `Grep`, `Glob` | purple |
| **Web** | `WebSearch`, `WebFetch` | teal |
| **Metadata** | `report_intent` | dim/ghost — minimal footprint |
| **Generic** | everything else (MCP tools, etc.) | blue (existing) |

### Card designs

All cards share the same outer container (`surface-card` with colored left border). The inner layout varies by category.

#### File Read

```
┌─────────────────────────────────────────────────┐
│ 📄 Read   src/middleware/auth.ts                │
├─────────────────────────────────────────────────┤
│ (collapsed by default — click to expand)        │
│                                                 │
│  1 │ import jwt from "jsonwebtoken";            │
│  2 │ import type { Request } from "express";    │
│  3 │ ...                                        │
│  [showing 3 of 47 lines — click to expand]      │
└─────────────────────────────────────────────────┘
```

- **Header**: icon + tool name + **basename** of file path (full path in title tooltip)
- **Body (collapsed)**: first 3 lines of file content as syntax-highlighted preview, line-numbered
- **Body (expanded)**: full content, scrollable, max-height 400px
- If result is not yet available (unpaired tool_use), show just the header with a subtle spinner/pulse

#### File Edit

```
┌─────────────────────────────────────────────────┐
│ ✏️ Edit   auth.ts  (+6 −2)                      │
├─────────────────────────────────────────────────┤
│  − const secret = "hardcoded";                  │
│  + const secret = process.env.JWT_SECRET;       │
│  + if (!secret) throw new Error("missing");     │
│  [3 more changes — click to expand]             │
└─────────────────────────────────────────────────┘
```

- **Header**: icon + tool name + basename + `(+N −M)` line count badge
- **Body**: unified diff with red/green line highlighting
- **Diff source** (in priority order):
  1. Parse `detailedContent` from tool_result if it contains a unified diff (Copilot emits this)
  2. Construct a minimal diff from `old_string` → `new_string` in the tool_use args (Claude Code path)
  3. Fall back to showing the raw args
- **Collapsed**: first 5 diff lines; **Expanded**: full diff

#### File Write

```
┌─────────────────────────────────────────────────┐
│ 📝 Write   config.ts  (new file, 24 lines)     │
├─────────────────────────────────────────────────┤
│  1 │ export const config = {                    │
│  2 │   port: 3000,                              │
│  3 │ ...                                        │
└─────────────────────────────────────────────────┘
```

- Like File Read but with green accent and "new file" indicator
- Shows content preview if available in args

#### Shell Command

```
┌─────────────────────────────────────────────────┐
│ $ npm install jsonwebtoken          ✓ exit 0    │
├─────────────────────────────────────────────────┤
│ added 1 package in 2.3s                         │
│ [2 more lines — click to expand]                │
└─────────────────────────────────────────────────┘
```

- **Header**: `$` prompt + command text (extracted from args, not JSON) + exit code badge on right
  - Exit 0: green `✓ exit 0`; nonzero: red `✗ exit 1`
  - If no exit code available yet (in-progress): pulsing dot
- **Command extraction**: strip JSON wrapper, unescape paths, show just the meaningful command
  - `"\"C:\\\\Program Files\\\\PowerShell\\\\7\\\\pwsh.exe\" -Command 'Get-Content -Path README.md'"` → `Get-Content -Path README.md` (extract the inner command when it's a pwsh wrapper)
  - Simple cases: just show `args.command` directly
- **Body**: terminal-style monospace output (dark inset background)
- **Collapsed**: first 3 lines; **Expanded**: full output

#### Search (Grep/Glob)

```
┌─────────────────────────────────────────────────┐
│ 🔍 Grep   "verifyToken"  in src/               │
├─────────────────────────────────────────────────┤
│ src/middleware/auth.ts:14                        │
│ src/routes/protected.ts:8                        │
│ [3 more matches]                                │
└─────────────────────────────────────────────────┘
```

- **Header**: icon + tool + pattern (in quotes) + search path
- **Body**: file match list or content matches

#### Metadata (report_intent, etc.)

```
│ ↳ intent: Updating README getting started       │
```

- Single inline line, no card border, dimmed text — treated like a system annotation, not an action
- Takes up minimal vertical space

#### Generic / MCP

```
┌─────────────────────────────────────────────────┐
│ ⚙ mcp__github__create_pr   #142                │
├─────────────────────────────────────────────────┤
│ { "title": "Add auth", "base": "main" }         │
│ ✓ PR created: #142                              │
└─────────────────────────────────────────────────┘
```

- Uses existing `argsPreview()` logic for header summary
- Formatted (indented) JSON for args, not raw escaped strings
- For MCP tools: strip `mcp__server__` prefix, show as `server / tool_name`

---

### Current state (`EventRenderer.tsx`)

```
EventRenderer
├─ ToolUseEvent     — unpaired tool_use (Claude Code path)
│   renders: > ToolName + JSON.stringify(args)
│
└─ ToolResultEvent  — paired tool_result (Copilot/Codex path)
    renders: ✓ Tool output + raw content <pre>
```

Two code paths, same ugly output.

### Proposed state

```
EventRenderer
└─ ToolCard          — unified component for both paired and unpaired
    ├─ classifyTool(toolName) → category
    └─ switch(category)
        ├─ FileReadCard
        ├─ FileEditCard
        ├─ FileWriteCard
        ├─ ShellCard
        ├─ SearchCard
        ├─ MetadataCard (inline, no card wrapper)
        └─ GenericToolCard
```

**Key changes:**

1. **Unified `ToolCard`** replaces both `ToolUseEvent` and `ToolResultEvent`
   - Receives: `tool: string`, `args: unknown`, `result?: string`, `isError?: boolean`, `detailedResult?: string`
   - When result is missing (unpaired/in-progress), shows header-only with loading indicator

2. **`classifyTool(name: string): ToolCategory`** — pure function, maps tool names to categories
   - Must handle cross-runtime synonyms (Read/view, Edit/edit, Bash/command_execution)

3. **Diff renderer** — lightweight component for edit cards
   - Parses unified diff format (lines starting with `+`/`-`/`@@`)
   - OR constructs diff from old_string/new_string
   - Red/green line backgrounds, no full diff library needed (these are small diffs)

4. **`parseToolResult(tool, category, content, detailedContent)`** — extracts structured data from raw result strings
   - Shell: parse `[exit N] output` format
   - Edit: extract diff from `detailedContent` JSON field
   - Read: pass through as file content

### Pairing changes

The existing `pairToolEvents()` in `sessionEvents.ts` stays the same but the `DisplayEvent` type gets richer:

```typescript
export type DisplayEvent = SessionEvent & {
  toolUseCtx?: {
    tool: string;
    args: unknown;
    /** Parsed from tool_result raw field, if available */
    detailedResult?: string;
  };
};
```

The `detailedResult` is extracted from the tool_result's content when it's a JSON object with `detailedContent` (Copilot's format). This lets the ToolCard access the diff without re-parsing.

---

## Mock data updates

Add a fourth scenario to `MOCK_STREAM_SCENARIOS` that exercises tool cards. This is what the demo site will show:

### Scenario D — "Fix Auth Bug" (tool-heavy)

```
[status: running]
[text: "Let me investigate the authentication bug..."]
[tool_use: Read src/middleware/auth.ts]
[tool_result: file contents (47 lines)]
[text: "Found the issue — the token expiry check is using `<` instead of `>`. Let me fix it."]
[tool_use: Edit src/middleware/auth.ts (old: "token.exp < now", new: "token.exp > now")]
[tool_result: success + diff]
[tool_use: Bash "npm test -- --grep auth"]
[tool_result: exit 0, "14 tests passed, 0 failed"]
[tool_use: Grep "token.exp" in src/]
[tool_result: 2 file matches]
[text: "Fixed. The token expiry comparison was inverted. All auth tests pass now."]
[usage: 42.6k tokens · $0.12]
[status: completed]
```

This gives the demo site realistic tool card content. Each tool event needs proper `raw` fields with IDs so pairing works.

---

## Scope

- **In scope**: `EventRenderer.tsx`, `EventRenderer.module.scss`, `sessionEvents.ts`, `mockData.ts`, Storybook stories
- **Out of scope**: server-side changes, proto changes, new data from runtimes — this is purely a rendering improvement using data we already have
- **No new dependencies**: diff rendering is simple enough to do inline (line-by-line `+`/`-` prefix parsing). No need for a diff library.

## Test plan

- [ ] Storybook stories for each tool card category (FileRead, FileEdit, Shell, Search, Generic)
- [ ] Storybook stories for edge cases: unpaired tool_use (in-progress), error results, very long output, empty output
- [ ] `?mock` mode shows tool cards in the streaming scenario
- [ ] E2E test: spawn a real session, verify tool cards render with correct structure (data-testid checks)

### PR DESCRIPTION
**Pull request by nick-pape**
## Summary

- Replace generic JSON blob rendering of tool_use/tool_result events with purpose-built card components: **FileReadCard** (line-numbered preview), **FileEditCard** (red/green diff), **ShellCard** (terminal output + exit code badge), **SearchCard** (pattern + match count), **MetadataCard** (inline annotation), **GenericToolCard** (MCP/unknown fallback)
- ToolCard router classifies tool names across runtimes (Claude Code, Copilot, Codex) and delegates to the appropriate sub-component — each independently testable via Storybook
- Add mock data Scenario D ("Fix Auth Bug") with tool_use/tool_result events for the `?mock` demo site

## Test plan
- [ ] Storybook renders all 31 stories across 7 card components
- [ ] `rush build` passes clean
- [ ] `?mock` mode streams tool cards in Scenario D
- [ ] Real sessions (claude-code, copilot, codex) render tool cards correctly

Closes #935

### CODE REVIEW

File: packages/web/src/components/tools/ShellCard.tsx [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+  const parsed = result ? parseShellOutput(result) : null;
+  const outputLines = parsed?.output.split("\n") ?? [];
+  const hasMore = outputLines.length > PREVIEW_LINES;
+  const displayLines = expanded ? outputLines : outputLines.slice(0, PREVIEW_LINES);
+  const exitCode = parsed?.exitCode ?? null;
+
+  return (
+    <div
+      className={`${styles.card} ${isError ? styles.cardRed : styles.cardNeutral}`}
```
[the remark this answers is not in the document]
Comment (nick-pape) [reply packages/web/src/components/tools/ShellCard.tsx]: Fixed — ShellCard now derives error state from `isError || (exitCode !== undefined && exitCode !== 0)` so non-zero exits consistently render red.

File: packages/web/src/mocks/mockData.ts [hunk 2]
```diff
@@ -1040,4 +1040,135 @@ export const MOCK_STREAM_SCENARIOS: MockStreamScenario[] = [
...
+        event: {
+          eventType: "tool_use",
+          timestamp: ts(3400),
+          content: JSON.stringify({
+            tool: "Edit",
+            args: {
+              file_path: "/src/middleware/auth.ts",
+              old_string: "    if (decoded.exp < Date.now() / 1000) {",
+              new_string: "    if (decoded.exp > Date.now() / 1000) {",
+            },
```
[the remark this answers is not in the document]
Comment (nick-pape) [reply packages/web/src/mocks/mockData.ts]: Fixed — swapped the bug direction. The mock code now has \ (incorrectly accepts expired tokens) and the fix changes it to \ (correct).

File: packages/web/src/components/tools/parseDiff.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+ */
+export function parseUnifiedDiff(diff: string): DiffLine[] {
+  const lines = diff.split("\n");
+  const result: DiffLine[] = [];
+
+  for (const line of lines) {
+    if (line.startsWith("diff --git") || line.startsWith("index ")) {
+      continue;
+    }
+    if (line.startsWith("---") || line.startsWith("+++")) {
```
[the remark this answers is not in the document]
Comment (nick-pape) [reply packages/web/src/components/tools/parseDiff.ts]: Fixed in earlier commit.
